debi.1
.TH DEBI 1 "Debian Utilities" "DEBIAN" \" -*- nroff -*-
.SH NAME
debi \- install current version of generated Debian package
.SH SYNOPSIS
\fBdebi\fP [\fIoptions\fR] [\fIchanges file\fR] [\fIpackage\fR ...]
.SH DESCRIPTION
\fBdebi\fR figures out the current version of a package and installs
it.  If a \fI.changes\fR file is specified on the command line, the
filename must end with \fI.changes\fR, as this is how the program
distinguishes it from package names.  If not, then \fBdebi\fR has to
be called from within the source code directory tree.  In this case,
it will look for the \fI.changes\fR file corresponding to the current
package version (by determining the name and version number from the
changelog, and the architecture in the same way as
\fBdpkg-buildpackage\fR(1) does).  It then runs \fBdebpkg -i\fR on
every \fI.deb\fR archive listed in the \fI.changes\fR file to install
them, assuming that all of the \fI.deb\fR archives live in the same
directory as the \fI.changes\fR file.  Note that you probably don't
want to run this program on a \fI.changes\fR file relating to a
different architecture after cross-compiling the package!
.PP
If a list of packages is given on the command line, then only those
debs with names in this list of packages will be installed.
.PP
Since installing a package requires root privileges, \fBdebi\fR calls
\fBdebpkg\fR rather than \fBdpkg\fR directly.  Thus \fBdebi\fR will
only be useful if it is either being run as root or \fBdebpkg\fR can
be run as root.  See \fBdebpkg\fR(1) for more details.
.SH "Directory name checking"
In common with several other scripts in the \fBdevscripts\fR package,
\fBdebi\fR will climb the directory tree until it finds a
\fIdebian/changelog\fR file.  As a safeguard against stray files
causing potential problems, it will examine the name of the parent
directory once it finds the \fIdebian/changelog\fR file, and check
that the directory name corresponds to the package name.  Precisely
how it does this is controlled by two configuration file variables
DEVSCRIPTS_CHECK_DIRNAME_LEVEL and DEVSCRIPTS_CHECK_DIRNAME_REGEX, and
their corresponding command-line options \fB\-\-check-dirname-level\fR
and \fB\-\-check-dirname-regex\fR.
.PP
DEVSCRIPTS_CHECK_DIRNAME_LEVEL can take the following values:
.TP
.B 0
Never check the directory name.
.TP
.B 1
Only check the directory name if we have had to change directory in
our search for \fIdebian/changelog\fR.  This is the default behaviour.
.TP
.B 2
Always check the directory name.
.PP
The directory name is checked by testing whether the current directory
name (as determined by \fBpwd\fR(1)) matches the regex given by the
configuration file option DEVSCRIPTS_CHECK_DIRNAME_REGEX or by the
command line option \fB\-\-check-dirname-regex\fR \fIregex\fR.  Here
\fIregex\fR is a Perl regex (see \fBperlre\fR(3perl)), which will be
anchored at the beginning and the end.  If \fIregex\fR contains a '/',
then it must match the full directory path.  If not, then it must
match the full directory name.  If \fIregex\fR contains the string
\'PACKAGE', this will be replaced by the source package name, as
determined from the changelog.  The default value for the regex is:
\'PACKAGE(-.*)?', thus matching directory names such as PACKAGE and
PACKAGE-version.
.SH OPTIONS
.TP
\fB\-a\fIdebian-architecture\fR, \fB\-t\fIGNU-system-type\fR
See \fBdpkg-architecture\fR(1) for a description of these options.
They affect the search for the \fI.changes\fR file.  They are provided
to mimic the behaviour of \fBdpkg-buildpackage\fR when determining the
name of the \fI.changes\fR file.
.TP
\fB\-\-debs\-dir\fR \fIDIR\fR
Look for the \fI.changes\fR and \fI.deb\fR files in directory
\fIDIR\fR instead of the parent of the source directory.  This should
either be an absolute path or relative to the top of the source
directory.
.TP
\fB\-\-check-dirname-level\fR \fIN\fR
See the above section "Directory name checking" for an explanation of
this option.
.TP
\fB\-\-check-dirname-regex\fR \fIregex\fR
See the above section "Directory name checking" for an explanation of
this option.
.TP
\fB\-\-no-conf\fR, \fB\-\-noconf\fR
Do not read any configuration files.  This can only be used as the
first option given on the command-line.
.TP
\fB\-\-help\fR, \fB\-\-version\fR
Show help message and version information respectively.
.SH "CONFIGURATION VARIABLES"
The two configuration files \fI/etc/devscripts.conf\fR and
\fI~/.devscripts\fR are sourced in that order to set configuration
variables.  Command line options can be used to override configuration
file settings.  Environment variable settings are ignored for this
purpose.  The currently recognised variables are:
.TP
.B DEBRELEASE_DEBS_DIR
This specifies the directory in which to look for the \fI.changes\fR
and \fI.deb\fR files, and is either an absolute path or relative to
the top of the source tree.  This corresponds to the
\fB\-\-debs\-dir\fR command line option.  This directive could be
used, for example, if you always use \fBpbuilder\fR or
\fBsvn-buildpackage\fR to build your packages.  Note that it also
affects \fBdebrelease\fR(1) in the same way, hence the strange name of
the option.
.TP
.BR DEVSCRIPTS_CHECK_DIRNAME_LEVEL ", " DEVSCRIPTS_CHECK_DIRNAME_REGEX
See the above section "Directory name checking" for an explanation of
these variables.  Note that these are package-wide configuration
variables, and will therefore affect all \fBdevscripts\fR scripts
which check their value, as described in their respective manpages and
in \fBdevscripts.conf\fR(5).
.SH "SEE ALSO"
.BR debpkg (1)
and
.BR devscripts.conf (5).
.SH AUTHOR
\fBdebi\fR was originally written by Christoph Lameter
<clameter@debian.org>.  The now-defunct script \fBdebit\fR was
originally written by James R. Van Zandt <jrv@vanzandt.mv.com>.  They
have been moulded into one script together with \fBdebc\fR(1) and
parts extensively modified by Julian Gilbey <jdg@debian.org>.


debi.pl
#! /usr/bin/perl -w

# debi:  Install current version of deb package
# debc:  List contents of current version of deb package
#
# debi and debc originally by Christoph Lameter <clameter@debian.org>
# Copyright Christoph Lameter <clameter@debian.org>
# The now defunct debit originally by Jim Van Zandt <jrv@vanzandt.mv.com>
# Copyright 1999 Jim Van Zandt <jrv@vanzandt.mv.com>
# Modifications by Julian Gilbey <jdg@debian.org>, 1999-2003
# Copyright 1999-2003, Julian Gilbey <jdg@debian.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

use 5.008;
use strict;
use Getopt::Long;
use File::Basename;
use filetest 'access';
use Cwd;

my $progname = basename($0,'.pl');  # the '.pl' is for when we're debugging
my $modified_conf_msg;

sub usage_i {
    print <<"EOF";
Usage: $progname [options] [.changes file] [package ...]
  Install the .deb file(s) just created, as listed in the generated
  .changes file or the .changes file specified.  If packages are listed,
  only install those specified packages from the .changes file.
  Options:
    --no-conf or      Don\'t read devscripts config files;
      --noconf          must be the first option given
    -a<arch>          Search for .changes file made for Debian build <arch>
    -t<target>        Search for .changes file made for GNU <target> arch
    --debs-dir DIR    Look for the changes and debs files in DIR instead of
                      the parent of the current package directory
    --check-dirname-level N
                      How much to check directory names:
                      N=0   never
                      N=1   only if program changes directory (default)
                      N=2   always
    --check-dirname-regex REGEX
                      What constitutes a matching directory name; REGEX is
                      a Perl regular expression; the string \`PACKAGE\' will
                      be replaced by the package name; see manpage for details
                      (default: 'PACKAGE(-.*)?')
    --help            Show this message
    --version         Show version and copyright information

Default settings modified by devscripts configuration files:
$modified_conf_msg
EOF
}

sub usage_c {
    print <<"EOF";
Usage: $progname [options] [.changes file] [package ...]
  Display the contents of the .deb file(s) just created, as listed
  in the generated .changes file or the .changes file specified.
  If packages are listed, only display those specified packages
  from the .changes file.  Options:
    --no-conf or      Don\'t read devscripts config files;
      --noconf          must be the first option given
    -a<arch>          Search for changes file made for Debian build <arch>
    -t<target>        Search for changes file made for GNU <target> arch
    --debs-dir DIR    Look for the changes and debs files in DIR instead of
                      the parent of the current package directory
    --check-dirname-level N
                      How much to check directory names:
                      N=0   never
                      N=1   only if program changes directory (default)
                      N=2   always
    --check-dirname-regex REGEX
                      What constitutes a matching directory name; REGEX is
                      a Perl regular expression; the string \`PACKAGE\' will
                      be replaced by the package name; see manpage for details
                      (default: 'PACKAGE(-.*)?')
    --help            Show this message
    --version         Show version and copyright information

Default settings modified by devscripts configuration files:
$modified_conf_msg
EOF
}

if ($progname eq 'debi') { *usage = \&usage_i; }
elsif ($progname eq 'debc') { *usage = \&usage_c; }
else { die "Unrecognised invocation name: $progname\n"; }

my $version = <<"EOF";
This is $progname, from the Debian devscripts package, version ###VERSION###
This code is copyright 1999-2003, Julian Gilbey <jdg\@debian.org>,
all rights reserved.
Based on original code by Christoph Lameter and James R. Van Zandt.
This program comes with ABSOLUTELY NO WARRANTY.
You are free to redistribute this code under the terms of
the GNU General Public License, version 2 or later.
EOF

# Start by setting default values
my $debsdir = '..';
my $debsdir_warning;
my $check_dirname_level = 1;
my $check_dirname_regex = 'PACKAGE(-.*)?';

# Next, read configuration files and then command line
# The next stuff is boilerplate

if (@ARGV and $ARGV[0] =~ /^--no-?conf$/) {
    $modified_conf_msg = "  (no configuration files read)";
    shift;
} else {
    my @config_files = ('/etc/devscripts.conf', '~/.devscripts');
    my %config_vars = (
		       'DEBRELEASE_DEBS_DIR' => '..',
		       'DEVSCRIPTS_CHECK_DIRNAME_LEVEL' => 1,
		       'DEVSCRIPTS_CHECK_DIRNAME_REGEX' => 'PACKAGE(-.*)?',
		       );
    my %config_default = %config_vars;

    my $shell_cmd;
    # Set defaults
    foreach my $var (keys %config_vars) {
	$shell_cmd .= qq[$var="$config_vars{$var}";\n];
    }
    $shell_cmd .= 'for file in ' . join(" ",@config_files) . "; do\n";
    $shell_cmd .= '[ -f $file ] && . $file; done;' . "\n";
    # Read back values
    foreach my $var (keys %config_vars) { $shell_cmd .= "echo \$$var;\n" }
    my $shell_out = `/bin/bash -c '$shell_cmd'`;
    @config_vars{keys %config_vars} = split /\n/, $shell_out, -1;

    # Check validity
    $config_vars{'DEVSCRIPTS_CHECK_DIRNAME_LEVEL'} =~ /^[012]$/
	or $config_vars{'DEVSCRIPTS_CHECK_DIRNAME_LEVEL'}=1;
    # We do not replace this with a default directory to avoid accidentally
    # installing a broken package
    $config_vars{'DEBRELEASE_DEBS_DIR'} =~ s%/+%/%;
    $config_vars{'DEBRELEASE_DEBS_DIR'} =~ s%(.)/$%$1%;
    if (! -d $config_vars{'DEBRELEASE_DEBS_DIR'}) {
	$debsdir_warning = "config file specified DEBRELEASE_DEBS_DIR directory $config_vars{'DEBRELEASE_DEBS_DIR'} does not exist!";
    }

    foreach my $var (sort keys %config_vars) {
	if ($config_vars{$var} ne $config_default{$var}) {
	    $modified_conf_msg .= "  $var=$config_vars{$var}\n";
	}
    }
    $modified_conf_msg ||= "  (none)\n";
    chomp $modified_conf_msg;

    $debsdir = $config_vars{'DEBRELEASE_DEBS_DIR'};
    $check_dirname_level = $config_vars{'DEVSCRIPTS_CHECK_DIRNAME_LEVEL'};
    $check_dirname_regex = $config_vars{'DEVSCRIPTS_CHECK_DIRNAME_REGEX'};
}

# Command line options next
my ($opt_help, $opt_version, $opt_a, $opt_t, $opt_debsdir);
my ($opt_ignore, $opt_level, $opt_regex, $opt_noconf);
GetOptions("help" => \$opt_help,
	   "version" => \$opt_version,
	   "a=s" => \$opt_a,
	   "t=s" => \$opt_t,
	   "debs-dir=s" => \$opt_debsdir,
	   "ignore-dirname" => \$opt_ignore,
	   "check-dirname-level=s" => \$opt_level,
	   "check-dirname-regex=s" => \$opt_regex,
	   "noconf" => \$opt_noconf,
	   "no-conf" => \$opt_noconf,
	   )
    or die "Usage: $progname [options] [.changes file] [package ...]\nRun $progname --help for more details\n";

if ($opt_help) { usage(); exit 0; }
if ($opt_version) { print $version; exit 0; }
if ($opt_noconf) {
    die "$progname: --no-conf is only acceptable as the first command-line option!\n";
}

my ($targetarch, $targetgnusystem);
$targetarch = $opt_a ? "-a$opt_a" : "";
$targetgnusystem = $opt_t ? "-t$opt_t" : "";

if ($opt_debsdir) {
    $opt_debsdir =~ s%/+%/%;
    $opt_debsdir =~ s%(.)/$%$1%;
    if (! -d $opt_debsdir) {
	$debsdir_warning = "--debs-dir directory $opt_debsdir does not exist!";
    }
    $debsdir = $opt_debsdir;
}

if ($debsdir_warning) {
    die "$progname: $debsdir_warning\n";
}

# dirname stuff
if ($opt_ignore) {
    die "$progname: --ignore-dirname has been replaced by --check-dirname-level and\n--check-dirname-regex; run $progname --help for more details\n";
}

if (defined $opt_level) {
    if ($opt_level =~ /^[012]$/) { $check_dirname_level = $opt_level; }
    else {
	die "$progname: unrecognised --check-dirname-level value (allowed are 0,1,2)\n";
    }
}

if (defined $opt_regex) { $check_dirname_regex = $opt_regex; }

# Is a .changes file listed on the command line?
my $changes;
if (@ARGV and $ARGV[0] =~ /\.changes$/) {
    $changes = shift;
}

my $chdir = 0;

if (! defined $changes) {
    # Look for .changes file via debian/changelog
    until (-r 'debian/changelog') {
	$chdir = 1;
	chdir '..' or die "$progname: can't chdir ..: $!\n";
	if (cwd() eq '/') {
	    die "$progname: cannot find readable debian/changelog anywhere!\nAre you in the source code tree?\n";
	}
    }

    # Find the source package name and version number
    my %changelog;
    open PARSED, q[dpkg-parsechangelog | grep '^\(Source\|Version\):' |]
	or die "$progname; cannot execute dpkg-parsechangelog | grep: $!\n";
    while (<PARSED>) {
	chomp;
	if (/^(\S+):\s(.+?)\s*$/) { $changelog{$1}=$2; }
	else {
	    die "$progname: don't understand dpkg-parsechangelog output: $_\n";
	}
    }

    close PARSED
	or die "$progname: problem executing dpkg-parsechangelog | grep: $!\n";
    if ($?) { die "$progname: dpkg-parsechangelog | grep failed!\n" }

    die "$progname: no package name in changelog!\n"
	unless exists $changelog{'Source'};
    die "$progname: no package name in changelog!\n"
	unless exists $changelog{'Version'};

    # Is the directory name acceptable?
    if ($check_dirname_level ==  2 or
	    ($check_dirname_level == 1 and $chdir)) {
	my $re = $check_dirname_regex;
	$re =~ s/PACKAGE/\\Q$changelog{'Source'}\\E/g;
	my $gooddir;
	if ($re =~ m%/%) { $gooddir = eval "cwd() =~ /^$re\$/;"; }
	else { $gooddir = eval "basename(cwd()) =~ /^$re\$/;"; }

	if (! $gooddir) {
	    my $pwd = cwd();
	    die <<"EOF";
$progname: found debian/changelog for package $changelog{'Source'} in the directory
  $pwd
but this directory name does not match the package name according to the
regex  $check_dirname_regex.

To run $progname on this package, see the --check-dirname-level and
--check-dirname-regex options; run $progname --help for more info.
EOF
	}
    }

    my $arch =
	`dpkg-architecture $targetarch $targetgnusystem -qDEB_HOST_ARCH`;
    if ($? != 0 or ! $arch) {
	die "$progname: unable to determine build architecture.\n";
    }
    chomp $arch;

    my $sversion = $changelog{'Version'};
    $sversion =~ s/^\d+://;
    my $pva="$changelog{'Source'}_${sversion}_${arch}";
    $changes="$debsdir/$pva.changes";
}

chdir dirname($changes)
    or die "$progname: can't chdir to $changes directory: $!\n";
$changes = basename($changes);

if (! -r $changes) {
    die "$progname: can't read $changes!\n";
}

my @debs = ();
my %pkgs = map { $_ => 0 } @ARGV;
open CHANGES, $changes or die "$progname: problem opening $changes: $!\n";
my $infiles;
while (<CHANGES>) {
    last if $infiles and /^[^ ]/;
    /^Files:/ and $infiles=1, next;
    next unless $infiles;
    if (/ (\S*\.deb)$/) {
        my $deb = $1;
        $deb =~ /^([a-z0-9+\.-]+)_/ or warn "unrecognised .deb name: $deb\n";
        my $pkg = $1;
        if (@ARGV) {
            if (exists $pkgs{$pkg}) {
                push @debs, $deb;
                $pkgs{$pkg}++;
            }
        } else {
            push @debs, $deb;
        }
    }
}
close CHANGES;

if ($progname eq 'debi') {
    system('debpkg', '-i', @debs) == 0
	or die "$progname: debpkg -i failed\n";
} else {
    # $progname eq 'debc'
    foreach my $deb (@debs) {
	print "$deb\n";
	print '-' x length($deb), "\n";
	system('dpkg-deb', '-I', $deb) == 0
	    or die "$progname: dpkg-deb -I $deb failed\n";
	system('dpkg-deb', '-c', $deb) == 0
	    or die "$progname: dpkg-deb -c $deb failed\n";
	print "\n";
    }
}

# Now do a sanity check
if (@ARGV) {
    foreach my $pkg (keys %pkgs) {
        if ($pkgs{$pkg} == 0) {
            warn "$progname: package $pkg not found in $changes, ignoring\n";
        } elsif ($pkgs{$pkg} > 1) {
            warn "$progname: package $pkg found more than once in $changes, installing all\n";
        }
    }
}

exit 0;
